govcd/vm.go

```diff
@@ -164,53 +168,91 @@ func (vm *VM) ChangeCPUCountWithCore(virtualCpuCount int, coresPerSocket *int) (
+				ipField, ipIsSet := network["ip"]
+				if ipIsSet {
+					ipFieldString = ipField.(string)
+				}
+
+				switch {
+				// TODO v3.0 remove from here when deprecated `ip` and `network_name` attributes are removed
+				case ipIsSet && ipFieldString == "dhcp": // Deprecated ip="dhcp" mode
 					ipAllocationMode = types.IPAllocationModeDHCP
-				} else if ip == "allocated" || network["ip"].(string) == "allocated" {
+				case ipIsSet && ipFieldString == "allocated": // Deprecated ip="allocated" mode
 					ipAllocationMode = types.IPAllocationModePool
-				} else if ip == "none" || network["ip"].(string) == "none" {
+				case ipIsSet && ipFieldString == "none": // Deprecated ip="none" mode
 					ipAllocationMode = types.IPAllocationModeNone
-				} else if ip != "" {
-					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = ip
-				} else if network["ip"].(string) != "" {
+
+				// Deprecated ip="valid_ip" mode (currently it is hit by ip_allocation_mode=MANUAL as well)
+				case ipIsSet && net.ParseIP(ipFieldString) != nil:
 					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = network["ip"].(string)
-				} else if ip == "" {
+					ipAddress = ipFieldString
+				case ipIsSet && ipFieldString != "": // Deprecated ip="something_invalid" we default to DHCP. This is odd but backwards compatible.
 					ipAllocationMode = types.IPAllocationModeDHCP
-				}
+					// TODO v3.0 remove until here when deprecated `ip` and `network_name` attributes are removed
 
-				util.Logger.Printf("[DEBUG] Function ChangeNetworkConfig() for %s invoked", network["orgnetwork"])
+				case ipIsSet && net.ParseIP(ipFieldString) != nil && (network["ip_allocation_mode"].(string) == types.IPAllocationModeManual):
+					ipAllocationMode = types.IPAllocationModeManual
+					ipAddress = ipFieldString
+				default: // New networks functionality. IP was not set and we're defaulting to provided ip_allocation_mode (only manual requires the IP)
+					ipAllocationMode = network["ip_allocation_mode"].(string)
+				}
 
-				networkSection.NetworkConnection[index].NeedsCustomization = true
-				networkSection.NetworkConnection[index].IPAddress = ipAddress
-				networkSection.NetworkConnection[index].IPAddressAllocationMode = ipAllocationMode
+				networkSection.NetworkConnection[loopIndex].NeedsCustomization = true
+				networkSection.NetworkConnection[loopIndex].IsConnected = true
+				networkSection.NetworkConnection[loopIndex].IPAddress = ipAddress
+				networkSection.NetworkConnection[loopIndex].IPAddressAllocationMode = ipAllocationMode
+
+				// for IPAllocationModeNone we hardcode special network name used by vcd 'none'
+				if ipAllocationMode == types.IPAllocationModeNone {
+					networkSection.NetworkConnection[loopIndex].Network = types.NoneNetwork
+				} else {
+					// Network names can be stored in two forms 'name' or 'network_name'. We pick
```

Comment:
Sorry, but where are these two variants `name` and `network_name` coming from?

Reply:
They come in from client consuming `func (vm *VM) ChangeNetworkConfig(networks []map[string]interface{}) (Task, error)`. But you are right in this case - we're not even using this function now. Premature optimization probably....

Reply:
Removed